### PR TITLE
Feature task list

### PR DESCRIPTION
**Pull request by at-trungnguyen**

### CODE REVIEW

File: app/src/main/java/com/app/android/data/source/LocalRepository.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import android.content.Context
+import com.app.android.data.source.datasource.LocalDataSource
+import com.app.android.pref.Pref
+
+/**
+ *
+ * Created by tien.hoang on 12/1/17.
+ */
+class LocalRepository : LocalDataSource {
+
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/data/source/LocalRepository.kt]: This class is never used. Please remove it

File: app/src/main/java/com/app/android/data/source/datasource/LocalDataSource.kt [hunk 2]
```diff
@@ -4,4 +4,6 @@ package com.app.android.data.source.datasource
  * This class contain methods work with data on local
  * @author at-vinhhuynh
  */
-interface LocalDataSource
+interface LocalDataSource {
+
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/data/source/datasource/LocalDataSource.kt]: This class is never used. Please remove it

File: app/src/main/java/com/app/android/data/source/datasource/TaskDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * @author at-vinhhuynh
+ */
+interface TaskDataSource {
+
+    /**
+     * This method use to get list task
+     *
+     * @param key firebase api key
+     * @param requestBody body
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/data/source/datasource/TaskDataSource.kt]: This method doesn't have any param. Please update the `javadoc`

File: app/src/main/java/com/app/android/data/source/remote/TaskRemoteDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import com.app.android.data.source.remote.network.ApiService
+import io.reactivex.Observable
+
+/**
+ *
+ * @author at-vinhhuynh
+ */
+class TaskRemoteDataSource(private val apiService: ApiService) : TaskDataSource {
+
+    constructor() : this(ApiClient.getInstance(null).service)
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/data/source/remote/TaskRemoteDataSource.kt]: The `default constructor` can be write in the `class header`. Like this:
`class TaskRemoteDataSource(private val apiService: ApiService = ApiClient.getInstance(null).service) : TaskDataSource {
`

Comment (at-tuannguyen) [reply app/src/main/java/com/app/android/data/source/remote/TaskRemoteDataSource.kt]: `ApiClient.getInstance(null).service`. This code is very ugly(:D). Could you please go to the `ApiClient` and modify:
`fun getInstance(arg: A?)` -> `fun getInstance(arg: A? = null)`

File: app/src/main/java/com/app/android/data/source/remote/network/ApiClient.kt [hunk 5]
```diff
@@ -23,6 +25,9 @@ open class ApiClient private constructor(url: String? = null) {
 
     private fun createService(): ApiService {
         val httpClientBuilder = OkHttpClient.Builder()
+        if (BuildConfig.DEBUG) {
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/data/source/remote/network/ApiClient.kt]: 👍

File: app/src/main/java/com/app/android/data/source/remote/network/ApiService.kt [hunk 6]
```diff
@@ -1,19 +1,18 @@
...
  *
  * @author at-vinhhuynh
  */
 interface ApiService {
+
     /**
-     * This method get github profile
-     *
-     * @param none
+     * This method use to create new user
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/data/source/remote/network/ApiService.kt]: The description is wrong. Please update

File: app/src/main/java/com/app/android/data/source/remote/network/ApiService.kt [hunk 7]
```diff
@@ -1,19 +1,18 @@
...
      */
-    @GET("user?access_token=???")
-    fun getProfile(): Single<LoginResponse>
+    @GET("/api/tasks")
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/data/source/remote/network/ApiService.kt]: By updating the `base URL` like this `http://172.17.29.7:8080``/`. You don't have to add `/` every API

File: app/src/main/java/com/app/android/ui/main/MainActivityUI.kt [hunk 8]
```diff
@@ -1,17 +1,67 @@
...
+
+    internal lateinit var swipeRefreshLayout: SwipeRefreshLayout
+    internal val taskListAdapter = TaskListAdapter(tasks)
+    internal lateinit var progressBar: ProgressBar
+
     override fun createView(ui: AnkoContext<MainActivity>) = with(ui) {
         relativeLayout {
-            
+            lparams(matchParent, matchParent)
+            backgroundColor = ContextCompat.getColor(ctx, android.R.color.white)
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/ui/main/MainActivityUI.kt]: Short way: `backgroundColorResource = Color.WHITE`

File: app/src/main/java/com/app/android/ui/main/MainActivityUI.kt [hunk 9]
```diff
@@ -1,17 +1,67 @@
...
+
+            toolbar {
+                id = R.id.main_activity_tool_bar
+                title = resources.getString(R.string.main_activity_title)
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/ui/main/MainActivityUI.kt]: Short way: `titleResource = R.string.main_activity_title`

File: app/src/main/java/com/app/android/ui/main/MainActivityUI.kt [hunk 10]
```diff
@@ -1,17 +1,67 @@
...
 
 /**
  *
  * @author at-vinhhuynh
  */
-class MainActivityUI : AnkoComponent<MainActivity> {
+class MainActivityUI(tasks: MutableList<Task>) : AnkoComponent<MainActivity> {
+
+    internal lateinit var swipeRefreshLayout: SwipeRefreshLayout
+    internal val taskListAdapter = TaskListAdapter(tasks)
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/ui/main/MainActivityUI.kt]: Please move it to line #25

File: app/src/main/java/com/app/android/ui/main/TaskListAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+            itemView.onClick {
+                onItemClick(tasks[adapterPosition])
+            }
+        }
+
+        /**
+         * Bind data to view holder.
+         */
+        fun onBind() {
+            with(tasks[adapterPosition]) {
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/ui/main/TaskListAdapter.kt]: Very expensive to get `adapterPosition `, please browse the function for more detail. Another hand, `position` can get from `onBindViewHolder `, use this way instead

File: app/src/main/java/com/app/android/ui/main/TaskListAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+                tvTitle.text = title
+                tvDescription.text = description
+                tvStatus.run {
+                    if (isDone == 1) {
+                        text = resources.getString(R.string.task_done)
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/ui/main/TaskListAdapter.kt]: Use `textResource` instead

File: app/src/main/java/com/app/android/ui/main/TaskListAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+                        textColor = ContextCompat.getColor(context, android.R.color.holo_green_light)
+                    } else {
+                        text = resources.getString(R.string.task_doing)
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/ui/main/TaskListAdapter.kt]: Use `textResource` instead

File: app/src/main/java/com/app/android/ui/main/TaskListAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        override fun createView(ui: AnkoContext<ViewGroup>): View {
+            val view = with(ui) {
+                verticalLayout {
+                    lparams(wrapContent, wrapContent)
+                    padding = dimen(R.dimen.item_task_padding)
+
+                    textView {
+                        id = R.id.item_task_title
+                        maxLines = 1
+                        textColor = ContextCompat.getColor(ctx, android.R.color.black)
```
Comment (at-tuannguyen) [inline app/src/main/java/com/app/android/ui/main/TaskListAdapter.kt]: With some basic colors, you can use this way: `textColorResource = Color.BLACK`

File: app/src/main/res/menu/main_menu.xml [hunk 15]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:android="http://schemas.android.com/apk/res/android">
+    <item
+        android:id="@+id/add_task"
+        android:title="@string/menu_add_task" />
```
Comment (at-tuannguyen) [inline app/src/main/res/menu/main_menu.xml]: Please add
```
        android:icon="@drawable/ic_add_black_24dp"
        app:showAsAction="always"
```

**Comment by at-trungnguyen** [comment]
@at-tuannguyen I fixed, please check again!

**Pull request merged**